HPCC-29914 Add embedded wasm support

GordonSmith opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [x] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

GordonSmith committed:
HPCC-29914 Add embedded wasm support

Signed-off-by: Gordon Smith <GordonJSmith@gmail.com>

ghalliday commented on plugins/wasmembed/secure-enclave/secure-enclave.cpp:
```diff
@@ -0,0 +1,502 @@
+    {
+        TRACE("getUnsignedResult");
+        if (results[0].kind() == wasmtime::ValKind::I64)
+            return (int32_t)results[0].i64();
```
ditto

ghalliday commented on plugins/wasmembed/secure-enclave/secure-enclave.cpp:
```diff
@@ -0,0 +1,502 @@
+        TRACE("getUnsignedResult");
+        if (results[0].kind() == wasmtime::ValKind::I64)
+            return (int32_t)results[0].i64();
+        return results[0].i32();
```
probably best to cast to unsigned first - otherwise it is not clear if it is signed extended

ghalliday reviewed (changes requested):
@GordonSmith some initial comments.  A few general themes

i) Need to be careful about types, and integer overflow.
ii) unicode is tricky - especially sizes of utf8 strings.  I didn't check all the code, but some looks wrong.
iii) I didn't see any thread protection.  E.g., hash table access/modification and other items need to be thread safe.  I didn't check if there were any problems.

GordonSmith commented on plugins/wasmembed/secure-enclave/abi.hpp:
```diff
@@ -0,0 +1,157 @@
+    return result;
+}
+
+std::string global_encoding = "utf8";
```
It is constant for now - but will be configurable in a future version of wasmtime

GordonSmith commented on plugins/wasmembed/secure-enclave/abi.hpp:
```diff
@@ -0,0 +1,157 @@
+    if (global_encoding.compare("utf8") == 0)
+    {
+        alignment = 1;
+        byte_length = tagged_code_units;
```
They do distinguish between tagged_code_points and tagged_code_units, so this should be good.

GordonSmith commented on plugins/wasmembed/secure-enclave/abi.hpp:
```diff
@@ -0,0 +1,157 @@
+void store_int(const wasmtime::Span<uint8_t> &data, int64_t v, size_t ptr, size_t nbytes, bool _signed = false)
+{
+    // convert v to little-endian byte array
+    std::vector<uint8_t> bytes(nbytes);
```
Not actually used - removing.

GordonSmith commented on plugins/wasmembed/secure-enclave/abi.hpp:
```diff
@@ -0,0 +1,157 @@
+    return input.substr(firstQuote + 1, secondQuote - firstQuote - 1);
+}
+
+std::pair<std::string, std::string> splitQualifiedID(const std::string &qualifiedName)
```
Refactored

GordonSmith commented:
@ghalliday I have a question WRT to threading and calling the actual function, I have two options:
1. Create a mutex around the actual call.
2. Make the "store" to a thread local (this will effectively ensure a separate instance of wasm "memory / state" per thread, and then no locks will be needed)

GordonSmith commented:
@ghalliday I have a question WRT to threading and calling the actual function, I have two options:
1. Create a mutex around the actual call.
2. Make the "store" to a thread local (this will effectively ensure a separate instance 

2 - makes more sense IMO

ghalliday commented on plugins/wasmembed/secure-enclave/secure-enclave.cpp:
```diff
@@ -0,0 +1,570 @@
+        uint32_t strPtr;
+        uint32_t code_units;
+        std::tie(strPtr, code_units) = load_string(data, ptr);
+        __chars = rtlUtf8Length(code_units, &data[strPtr]);
```
__chars is not a standard name for the result parameter

ghalliday commented on plugins/wasmembed/wasmembed.cpp:
```diff
@@ -0,0 +1,149 @@
+        }
+        void manifestPaths(ICodeContext *codeCtx)
+        {
+            if (codeCtx && !manifestAdded)
```
Does this need to be thread safe?

GordonSmith replied:
Nope - refactored away

ghalliday commented on plugins/wasmembed/wasmembed.cpp:
```diff
@@ -0,0 +1,149 @@
+
+MODULE_EXIT()
+{
+    kill();
```
This could do with a clearer name - hard to tell from the context what it does.  (I didn't search to find it.)

ghalliday reviewed (changes requested):
Various random comments.  A few comments from before are still relevant.

GordonSmith commented on plugins/wasmembed/secure-enclave/abi.cpp:
```diff
@@ -0,0 +1,138 @@
+    uint8_t b = data[ptr + i];
+    if (i == nbytes - 1 && std::is_signed<T>::value && b >= 0x80)
+      b -= 0x100;
+    retVal += b << (i * 8);
```
b is a uint8_t?

GordonSmith commented on plugins/wasmembed/secure-enclave/abi.cpp:
```diff
@@ -0,0 +1,138 @@
+//  More:  Not currently available from the wasmtime::context object, see https://github.com/bytecodealliance/wasmtime/issues/6719
+std::string global_encoding = "utf8";
+
+std::pair<uint32_t /*ptr*/, uint32_t /*byte length*/> load_string_from_range(const wasmtime::Span<uint8_t> &data, uint32_t ptr, uint32_t tagged_code_units)
```
I will add a comment, but a "code unit" (and a "code point") are apparently the "correct" terminology to distinguish between a char and a byte:  https://unicode.org/glossary/

GordonSmith commented on plugins/wasmembed/secure-enclave/abi.cpp:
```diff
@@ -0,0 +1,138 @@
+
+int align_to(int ptr, int alignment)
+{
+  return std::ceil(ptr / alignment) * alignment;
```
Looking in https://github.com/WebAssembly/component-model/blob/5a11e36e43244130a71919b7cdd82139e5fe65b3/design/mvp/canonical-abi/definitions.py outside of the tests, it is used to find the next aligned ptr given a current ptr - so am inclined to leave as is for now and look at optimisation once the abi is complete?

GordonSmith commented on plugins/wasmembed/secure-enclave/secure-enclave.cpp:
```diff
@@ -0,0 +1,570 @@
+
+    ThreadSafeMap<std::string, wasmtime::Instance> wasmInstances;
+    //  wasmMems and wasmFuncs are only written to during createInstance, so no need for a mutex
+    std::unordered_map<std::string, wasmtime::Memory> wasmMems;
```
Updated comment, but yes its only called behind the ThreadSafeMap mutex

ghalliday commented on plugins/wasmembed/secure-enclave.cpp:
```diff
@@ -0,0 +1,677 @@
+        uint32_t strPtr;
+        uint32_t bytes;
+        std::tie(strPtr, bytes) = load_string(data, ptr);
+        rtlStrToStrX(chars, result, bytes, reinterpret_cast<const char *>(&data[strPtr]));
```
Need to convert from utf8 to a string.

ghalliday reviewed (commented):
@GordonSmith a few comments relating to c++ details, but broadly looks good.

GordonSmith commented on plugins/wasmembed/CMakeLists.txt:
```diff
@@ -0,0 +1,67 @@
+project(wasmembed)
+
+set(CMAKE_CXX_STANDARD 20)
```
It should only be local to the wasmembed project, but I can relocate anyway

GordonSmith commented on plugins/wasmembed/abi.cpp:
```diff
@@ -0,0 +1,268 @@
+
+uint32_t align_to(uint32_t ptr, uint32_t alignment)
+{
+    return std::ceil(ptr / alignment) * alignment;
```
I agree WRT to the naming, but want to keep the names in sync with the official ABI spec.

GordonSmith commented on testing/regress/ecl/wasmembed.ecl:
```diff
@@ -0,0 +1,87 @@
+s16Test(-32768, -1) = (integer2)(-32768 - 1);
+
+s32Test(1, 2) = (integer4)(1 + 2);
+s32Test(2147483647, 1) = (integer4)(2147483647 + 1);
```
The test was to ensure that they both behaved the same in this scenario - should I remove?

GordonSmith commented on plugins/wasmembed/wasmembed.cpp:
```diff
@@ -0,0 +1,74 @@
+
+    extern DECL_EXPORT IEmbedContext *getEmbedContext()
+    {
+        return LINK(&theEmbedContext);
```
I modelled it on:  https://github.com/hpcc-systems/HPCC-Platform/blob/master/plugins/v8embed/v8embed.cpp#L938-L982

GordonSmith commented on plugins/wasmembed/util.cpp:
```diff
@@ -0,0 +1,61 @@
+
+std::vector<uint8_t> readWasmBinaryToBuffer(const std::string &filename)
+{
+    std::ifstream file(filename, std::ios::binary | std::ios::ate);
```
(when I wrote it, it didn't have access to jlib)

GordonSmith commented on plugins/wasmembed/secure-enclave.cpp:
```diff
@@ -0,0 +1,677 @@
+
+        wasmtime::Linker linker(engine);
+        linker.define_wasi().unwrap();
+        TRACE("WASM SE resolveModule4 %s", wasmName.c_str());
```
Agree...

ghalliday commented on plugins/wasmembed/abi.cpp:
```diff
@@ -124,12 +129,13 @@ T load_int(const wasmtime::Span<uint8_t> &data, int32_t ptr)
     for (int i = 0; i < nbytes; ++i)
     {
-        uint8_t b = data[ptr + i];
+        std::make_unsigned_t<T> b = data[ptr + i];
```
retVal will also need to be unsigned, otherwise you can get integer overflow

ghalliday commented on plugins/wasmembed/abi.cpp:
```diff
@@ -124,12 +129,13 @@ T load_int(const wasmtime::Span<uint8_t> &data, int32_t ptr)
         retVal += b << (i * 8);
     }
-    if (std::is_signed<T>::value)
+    if (std::is_signed<T>::value && data[ptr + nbytes - 1] & 0x80)
```
There also needs to be a check that nbytes != sizeof(T), because value << 8 * sizeof(T) is undefined behaviour.

There should also be a comment
//if the top bit is set then sign extend the result by setting the top bits.

NOTE: Running in debug on linux will throw errors when undefined behaviour is spotted.

ghalliday commented on plugins/wasmembed/secure-enclave.cpp:
```diff
@@ -551,7 +542,7 @@ class SecureFunction : public CInterfaceOf<IEmbedFunctionContext>
         uint32_t bytes;
         std::tie(strPtr, bytes) = load_string(data, ptr);
-        rtlStrToStrX(chars, result, bytes, reinterpret_cast<const char *>(&data[strPtr]));
+        rtlUtf8ToStrX(chars, result, bytes, reinterpret_cast<const char *>(&data[strPtr]));
```
careful about difference between length/codepoints and size (I wish it wasn't this way!)
```
size32_t codepoints = rtlUtfLength(bytes, strPtr);
rtlUtf8ToStrX(chars, result, codePoints, reinterpret_cast<const char *>(&data[strPtr]));
```

ghalliday reviewed (commented):
@GordonSmith looks close.  A few final comments

ghalliday commented on plugins/wasmembed/secure-enclave.cpp:
```diff
@@ -26,97 +24,137 @@
+
+    void registerInstance(const std::string &wasmName)
+    {
+        TRACE("WASM SE registerInstance %s", wasmName.c_str());
```
Are there thread safety issues here?  Currently thread local, so no, but will that cause other issues?

GordonSmith replied:
I don't believe so, this PR reworked the engine / store threading specifically to match the threaded examples in the wasmtime repo (engine is a global and store / store context is per thread).

ghalliday commented on plugins/wasmembed/secure-enclave.cpp:
```diff
@@ -266,29 +267,22 @@ class WasmEngine
     }
 };
-static std::unique_ptr<WasmEngine> wasmEngine;
+thread_local std::unique_ptr<WasmStore> wasmStore = std::make_unique<WasmStore>();
```
Potential problems with code like this on osx and efficiency of thread_local variables with non-trivial constructors.

GordonSmith replied:
I am not sure how the lifetime of threading works within thor jobs - but I suspect this must be more efficient than a lock on the call.

GordonSmith replied:
Open new ticket for efficiency access to wasm store:  Move `std::make_unique<WasmStore>();` into secure function constructor (testing if it exists first).

ghalliday commented on plugins/wasmembed/secure-enclave.cpp:
```diff
@@ -538,20 +517,23 @@ class SecureFunction : public CInterfaceOf<IEmbedFunctionContext>
-        std::tie(strPtr, bytes) = load_string(data, ptr);
-        rtlUtf8ToStrX(chars, result, bytes, reinterpret_cast<const char *>(&data[strPtr]));
+        std::tie(strPtr, encoding, bytes) = load_string(data, ptr);
+        size32_t codepoints = rtlUtf8Length(bytes, &data[strPtr]);
```
Should encoding be being used?

GordonSmith replied:
Not currently as everything is always a utf8 - this is more to keep in sync with official ABI and in the future when other encodings are supported.

ghalliday commented on plugins/wasmembed/secure-enclave.cpp:
```diff
@@ -614,55 +597,27 @@ class SecureFunction : public CInterfaceOf<IEmbedFunctionContext>
     virtual void compileEmbeddedScript(size32_t lenChars, const char *_utf) override
     {
         TRACE("WASM SE compileEmbeddedScript");
```
Should this fail?

ghalliday reviewed (commented):
@GordonSmith I suspect this is close to ready to merge. I think it might be worth a walkthrough with me and someone else to look at and address any remaining issues.

GordonSmith commented on plugins/wasmembed/secure-enclave.cpp:
```diff
@@ -26,97 +24,137 @@
+            throw makeStringException(100, "Faile to get engine context");
+
+        StringArray manifestModules;
+        engine->getManifestFiles("wasm", manifestModules);
```
ref:  https://track.hpccsystems.com/browse/HPCC-30650

GordonSmith commented on plugins/wasmembed/secure-enclave.cpp:
```diff
@@ -0,0 +1,624 @@
+            inData += thisSize;
+        }
+        args.push_back(memIdx);
+        args.push_back(numElems);
```
Strictly should throw unsupported.

GordonSmith commented on testing/regress/ecl/wasmembed.ecl:
```diff
@@ -0,0 +1,112 @@
+d2 := project(choosen(d, sampleSize), t(LEFT));
+d3 := project(choosen(d, sampleSize, 5000), t(LEFT));
+d4 := project(choosen(d, sampleSize, 10001), t(LEFT));
+count(d2(passed=false)) = 0 AND count(d3(passed=false)) = 0 AND count(d4(passed=false)) = 0;
```
Open ticket to add threaded testing.  Use table to get the average of "passed" and then an unordered set of those results.
```
c1 := TABLE(d1(not passed), { unsigned count = COUNT(GROUP) {);
all := (+)(c1, c2, c3, UNORDERED);
SUM(all, count = 0)
```
Type of thing...

GordonSmith commented on plugins/wasmembed/CMakeLists.txt:
```diff
@@ -0,0 +1,66 @@
+            find_library(WASMTIME_LIB NAMES wasmtime
+                PATHS ${VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}
+            )
+        endif()
```
Not needed as file name already matches

GordonSmith commented on plugins/wasmembed/secure-enclave.cpp:
```diff
@@ -0,0 +1,624 @@
+        TRACE("WASM SE WasmEngine");
+    }
+
+    ~WasmEngine()
```
https://track.hpccsystems.com/browse/HPCC-30692

ghalliday reviewed (approved):
@GordonSmith please squash.  I'll then quickly scan and merge if I don't spot anything.

GordonSmith commented:
> @GordonSmith please squash. I'll then quickly scan and merge if I don't spot anything.

Squashed and rebased - will re-review sometime this week once the tests succeeed

GordonSmith commented:
@ghalliday - removed some obsolete files and should be ready for pull

GordonSmith commented:
@ghalliday rebased
